### PR TITLE
Qutrit Non-Parametric Ops #2 (TAdd, TSWAP)

### PR DESCRIPTION
**Pull request by mudit2812**
**Context:**
There are very few simulators available today that offer qudit simulation. One notable example is [Google's Cirq](https://quantumai.google/cirq/qudits), which includes support for qudit operations. However, the simulator is far from complete, and a key drawback is the lack of operations available, which creates a need for users to define operations themselves. Popular libraries such as Qiskit and QuTiP also include some qudit manipulation capabilities, but with few use cases. Beyond these libraries, qudit simulation solutions are quite limited. Moreover, qudit simulation is a feature that is already in demand, as seen in this [issue](https://github.com/PennyLaneAI/pennylane/issues/2190) on the PennyLane repository.

This PR builds upon the previous PR's that added qutrit functionality by implementing the non-parametric two-qutrit Ternary Add and Swap operations.

**Description of the Change:**
* Added `TAdd()` operation, which is the generalized alanog of the `CNOT` gate. Added `TSWAP()` operation to PennyLane to use with qutrit devices.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**
https://github.com/PennyLaneAI/pennylane/issues/2190

### CODE REVIEW

File: pennylane/devices/default_qutrit.py
Comment (glassnotes) [inline pennylane/devices/default_qutrit.py:193]:
```suggestion
        """Applies a controlled ternary add gate by slicing along the first axis specified in ``axes`` and
```

File: pennylane/devices/default_qutrit.py
Comment (glassnotes) [inline pennylane/devices/default_qutrit.py:194]:
```suggestion
        applying a TShift transformation along the second axis.
```

File: pennylane/devices/default_qutrit.py
Comment (glassnotes) [inline pennylane/devices/default_qutrit.py:225]: It looks like this function is identical to the `_apply_swap` method in the qubit devices. Would it make sense to pull that function in from `DefaultQubit` rather than redefining it here? One could just set `"TSWAP": self._apply_swap,` in the dictionary above, even. 

On that note, this is something to note for future generalizations to qudit devices. If SWAP has the same implementation regardless of dimension, maybe it doesn't even need a different name.

File: pennylane/ops/qutrit/non_parametric_ops.py
Comment (glassnotes) [inline pennylane/ops/qutrit/non_parametric_ops.py:213]: Even though the reference is provided, I think it would be helpful to show the action explicitly because it is hard to glean this from a 9 x 9 matrix. We'll just have to check that it renders correctly.

```suggestion
    `Yeh et al. (2022) <https://arxiv.org/abs/2204.00552>`_. 
    It performs the controlled :class:`~.TShift` operation, and sends
    :math:`\hbox{TAdd} \vert i \rangle \vert j \rangle = \vert i \rangle \vert i + j \rangle`, 
    where addition is taken modulo 3. The matrix representation is
```

File: pennylane/ops/qutrit/non_parametric_ops.py
Comment (glassnotes) [inline pennylane/ops/qutrit/non_parametric_ops.py:321]: Now that this is included, we could also set the `_controlled` attribute of the `TShift` operation to be `TAdd`, (as is done with `PauliX` and `CNOT` in the qubit case). This could be done, though, in a separate, smaller PR as it may require a number of new tests, and also may need to be done for other operations.

File: pennylane/devices/default_qutrit.py
Comment (glassnotes) [inline pennylane/devices/default_qutrit.py:223]: Nice - this looks like the logical analog of the `_apply_cnot` method in the qubit case.

File: tests/ops/qutrit/test_qutrit_non_parametric_ops.py
Comment (glassnotes) [inline tests/ops/qutrit/test_qutrit_non_parametric_ops.py:171]:
```suggestion
involution_ops = [  # ops that are their own inverses
    qml.TSWAP(wires=[0, 1]),
]
```

File: tests/devices/test_default_qutrit.py
Comment (glassnotes) [inline tests/devices/test_default_qutrit.py:75]: Are these tests that were missing from the previous PR? The ones that would have been removed for inclusion with `expand_matrix`?

File: tests/devices/test_default_qutrit.py
Comment (glassnotes) [inline tests/devices/test_default_qutrit.py:99]: Just to clarify - subspace is here because it will be needed in future tests with the subspace ops, correct?

File: tests/devices/test_default_qutrit.py
Comment (glassnotes) [inline tests/devices/test_default_qutrit.py:120]:
```suggestion
        """Tests that applying an inverse operation yields the expected output state for single wire
```

File: tests/devices/test_default_qutrit.py
Comment (glassnotes) [inline tests/devices/test_default_qutrit.py:135]: For all three of these, I recommend to add an additional test case with complex values and 1-2 more terms in the superposition just to ensure that things aren't coming out correct just due to some symmetry in the input.

File: tests/devices/test_default_qutrit.py
Comment (glassnotes) [inline tests/devices/test_default_qutrit.py:200]:
```suggestion
        """Tests that applying an inverse operation yields the expected output state for two wire
```

File: tests/devices/test_default_qutrit.py
Comment (glassnotes) [inline tests/devices/test_default_qutrit.py:953]: Recommend to make the reverse version a separate test, as is done for the `DefaultQubit` tests.

**Review by glassnotes — CHANGES_REQUESTED** [review]
Nice work @mudit2812, I have only minor suggestions, and I left a couple questions in the test files. :+1:

**Review by Qottmann — COMMENTED** [review]
Great work @mudit2812 this was fun to review! Some minor things on documentation and some curiosity questions.

File: pennylane/ops/qutrit/non_parametric_ops.py
Comment (Jaybsoni) [reply pennylane/ops/qutrit/non_parametric_ops.py:247]: Maybe it means `wire_order` perhaps?

File: pennylane/ops/qutrit/non_parametric_ops.py
Comment (Jaybsoni) [reply pennylane/ops/qutrit/non_parametric_ops.py:319]: This is a common trick we use in qubit devices as well. 

The way we track inverses of operations right now in PL is with the `self.inverse` property. So if one takes the inverse of say `RX`, we don't construct a new operation `RX_inv` but just update the inverse property for the `RX` class itself. This leads to an edge case when one asks for the adjoint of the inverse of `RX`. We can't simply set the inverse again, rather we need to flip the status of the inverse property to account for the case when it is already inverted. That is what is happening here.

Comment (Qottmann) [reply pennylane/ops/qutrit/non_parametric_ops.py:319]: Thanks for clarifying!

File: pennylane/ops/qutrit/non_parametric_ops.py
Comment (Qottmann) [inline pennylane/ops/qutrit/non_parametric_ops.py:257]: One thing that came to my mind afterwards:
Thinking for simplicity in terms of CNOT, the matrix representation can look different depending on whether the first or the second qubit acts as a control. Is this in some way being taken into account?
image

File: pennylane/ops/qutrit/non_parametric_ops.py
Comment (glassnotes) [reply pennylane/ops/qutrit/non_parametric_ops.py:247]: It is the canonical matrix you would write if you were considering how the operation acts on the computational basis states in order, in a mathematical context independent of any devices.

File: pennylane/ops/qutrit/non_parametric_ops.py
Comment (glassnotes) [reply pennylane/ops/qutrit/non_parametric_ops.py:257]: Indeed, this is related to the above comment about textbook matrix representation and wire order. `Operation.compute_matrix` gives only the canonical representation, but one can call `Operation.matrix` which takes a `wire_order` parameter (which works by first calling `compute_matrix` and then `expand_matrix` if necessary to adjust the ordering).

Incidentally, `expand_matrix` actually has to be adjust to work for qutrit operations, which is going to be taken care of in a separate PR.

File: pennylane/ops/qutrit/non_parametric_ops.py
Comment (glassnotes) [reply pennylane/ops/qutrit/non_parametric_ops.py:304]: @mudit2812 can correct me if I'm wrong but I believe `compute_eigvals` generally matches the order obtained from `np.linalg.eigvals` (which is why this function is getting called in the related test for comparison and not being hardcoded).

File: pennylane/ops/qutrit/non_parametric_ops.py
Comment (glassnotes) [reply pennylane/ops/qutrit/non_parametric_ops.py:328]: Yep that's correct!

File: pennylane/ops/qutrit/non_parametric_ops.py
Comment (glassnotes) [inline pennylane/ops/qutrit/non_parametric_ops.py:330]:
```suggestion
    The ternary swap operator.
    
    This operation is analogous to the qubit SWAP and acts on two-qutrit computational basis states
    according to :math:`TSWAP\vert i, j\rangle = \vert j, i \rangle`. Its matrix representation is

    .. math:: TSWAP = \begin{bmatrix}
```

File: pennylane/devices/default_qutrit.py
Comment (Qottmann) [inline pennylane/devices/default_qutrit.py:228]: Forgot to mention this in the last review, would be also good to have the simple definition $\texttt{TSWAP}|i, j\rangle=|j, i\rangle$ already here

File: pennylane/ops/qutrit/non_parametric_ops.py
Comment (Qottmann) [reply pennylane/ops/qutrit/non_parametric_ops.py:250]: Ah thanks, in combination with the comment below I understand what is meant now. I think this could be rewritten, most people reading this will probably not understand this from "it does not consider wires".

**Comment by Jaybsoni** [comment]
Hi @glassnotes and @mudit2812, 

This is a great addition and I really look forward to having this generalization of qudit systems in PL. While I haven't done an in-depth review, one question I had while exploring the code was how well its integrated with existing "[operator functions](https://docs.pennylane.ai/en/stable/introduction/operations.html)". These operator functions take existing pennylane operators and return either PL operators or some scalar/tensor value. 

One example of this is the top level `qml.matrix()` function. This top level method is heavily used to determine the matrix of any PL object (qnode, op, quantum func). It takes a keyword arg `wire_order`, which re-orders and potentially expands the matrix to fit a larger Hilbert space. 

image

This isn't currently compatible with qutrit ops because of the implicit assumption in `expand_matrix()` that expansion by m qubits can be accomplished by tensoring an identity of size (2^{m} x 2^{m}).

I suspect there are similar assumptions littering the code base when it comes to these operator functions. Is this something that these set of PRs will be addressing? Maybe a separate PR from this one.

**Comment by glassnotes** [comment]
> This isn't currently compatible with qutrit ops because of the implicit assumption in expand_matrix() that expansion by m qubits can be accomplished by tensoring an identity of size (2^{m} x 2^{m}).

@Jaybsoni yup we are aware of the changes that are needed in `expand_matrix` and planning to make that a separate (smaller) PR for the sake of readability during review.

File: pennylane/devices/default_qutrit.py
Comment (glassnotes) [inline pennylane/devices/default_qutrit.py:229]:
```suggestion
        specified axes. The ternary SWAP gate acts on the computational basis states like 
        :math:`\vert i, j\rangle \rightarrow \vert j, i \rangle`.
```

File: pennylane/ops/qutrit/non_parametric_ops.py
Comment (glassnotes) [reply pennylane/ops/qutrit/non_parametric_ops.py:250]: I do agree but I am going to leave as-is for now, as this is written the same way in every single operation and changing them all would be out of scope. It is something better updated in `operation.py` where `compute_matrix` is initially defined.

File: pennylane/ops/qutrit/non_parametric_ops.py
Comment (Jaybsoni) [inline pennylane/ops/qutrit/non_parametric_ops.py:251]: I think this line might also add to the confusion when it mentions a global wire order

File: tests/ops/qutrit/test_qutrit_non_parametric_ops.py
Comment (Jaybsoni) [inline tests/ops/qutrit/test_qutrit_non_parametric_ops.py:134]: Is the swap left the same because swap^-1 = swap?

File: tests/ops/qutrit/test_qutrit_non_parametric_ops.py
Comment (Jaybsoni) [inline tests/ops/qutrit/test_qutrit_non_parametric_ops.py:75]: There should be a similar test for `TSWAP` ?

**Review by Jaybsoni — COMMENTED** [review]
Great work so far @mudit2812 ! 

Just a few comments and suggestions. Once addressed I am happy to approve 👍🏼

File: tests/ops/qutrit/test_qutrit_non_parametric_ops.py
Comment (mudit2812) [reply tests/ops/qutrit/test_qutrit_non_parametric_ops.py:87]: Not really. Most of the other testing data has been written as lists of tuples, which is why I think having the test data inside a list is better in general. I also think it's a bit more readable for other tests where there might be lots of nesting.

File: pennylane/ops/qutrit/__init__.py
Comment (mudit2812) [reply pennylane/ops/qutrit/__init__.py:35]: `TSWAP` is in fact Hermitian. While it can be used as an observable, I don't see the benefit for the same reason that `SWAP` isn't an observable. I'm not sure how useful it is, and it's a bit more restrictive in its use as any circuit potentially using it as an observable would need to have multiple wires.

File: pennylane/ops/qutrit/non_parametric_ops.py
Comment (mudit2812) [reply pennylane/ops/qutrit/non_parametric_ops.py:216]: I think that's a good idea as well. With the incoming deadline of 0.26, I think it's a bit infeasible, but I think that it's going to be helpful for future releases and I'll keep it in mind.

File: tests/devices/test_default_qutrit.py
Comment (mudit2812) [reply tests/devices/test_default_qutrit.py:108]: This is not a fixture. Some of the future qutrit operations will use subspaces (i.e. will only apply to a subset of the qutrit basis states, and not do anything to the other remaining basis states), so I've written the test this way so that if the `subspace` parameter is not `None` (currently `subsapce` is `None` for all test cases), the `subspace` keyword argument is included in the operation definition. This avoids a lot of repetitive code in the test suite down the line.

File: tests/devices/test_default_qutrit.py
Comment (mudit2812) [reply tests/devices/test_default_qutrit.py:112]: There isn't any implicit casting, but I'm following the convention from `test_default_qubit.py`.

File: tests/ops/qutrit/test_qutrit_non_parametric_ops.py
Comment (mudit2812) [reply tests/ops/qutrit/test_qutrit_non_parametric_ops.py:134]: Yep

File: tests/ops/qutrit/test_qutrit_non_parametric_ops.py
Comment (mudit2812) [reply tests/ops/qutrit/test_qutrit_non_parametric_ops.py:75]: `TSWAP` doesn't have an `eigvals` method. If I'm being completely honest, I'm not entirely sure that `TAdd` or any of the other non-parametric operations even need an `eigvals` method. I'd love to get some input on this.

File: pennylane/devices/default_qutrit.py
Comment (mudit2812) [reply pennylane/devices/default_qutrit.py:227]: Thats a good point. However, I can't import `_apply_swap` from `DefaultQubit` as a `DefaultQubit` object would need to be initialized for that method to be usable, which doesn't make sense unless I just have `DefaultQutrit` inherit `DefaultQubit`.

File: pennylane/ops/qutrit/non_parametric_ops.py
Comment (mudit2812) [reply pennylane/ops/qutrit/non_parametric_ops.py:324]: This can't be done, as `TShift` always shifts by one index, while the number of times `TAdd` shifts the target qutrit depends on the state of the control qutrit.

File: tests/devices/test_default_qutrit.py
Comment (mudit2812) [reply tests/devices/test_default_qutrit.py:75]: No these aren't. All tests pertaining to `expand_matrix` would've been in `tests/test_operations.py`

File: tests/devices/test_default_qutrit.py
Comment (mudit2812) [reply tests/devices/test_default_qutrit.py:99]: That's correct.

File: tests/devices/test_default_qutrit.py
Comment (mudit2812) [reply tests/devices/test_default_qutrit.py]: I can't find this code block anymore. I might've updated the tests before the summer ended and can't recall.

File: pennylane/devices/default_qutrit.py
Comment (mudit2812) [reply pennylane/devices/default_qutrit.py]: I added it! See line 195

File: pennylane/ops/qutrit/non_parametric_ops.py
Comment (mudit2812) [reply pennylane/ops/qutrit/non_parametric_ops.py:251]: I do agree with you. However, as @glassnotes mentioned in another comment, this comment is used for the `compute_matrix` method of every other operation as well, which is why I'm hesitant to change it at the moment.

File: tests/devices/test_default_qutrit.py
Comment (glassnotes) [reply tests/devices/test_default_qutrit.py:135]: Resolving this comment as it looks like these have been added. Nice work :star:

File: tests/devices/test_default_qutrit.py
Comment (glassnotes) [inline tests/devices/test_default_qutrit.py:974]:
```suggestion
        applied wires are reversed."""
```

File: pennylane/ops/qutrit/non_parametric_ops.py
Comment (Jaybsoni) [reply pennylane/ops/qutrit/non_parametric_ops.py:251]: So if I understand correctly this is the same docstring as those for most of the other operations in PL? or in the qutrit project?

File: tests/ops/qutrit/test_qutrit_non_parametric_ops.py
Comment (glassnotes) [reply tests/ops/qutrit/test_qutrit_non_parametric_ops.py:75]: Just looking at the non-parametric qubit operations, I can't actually figure out why some of the operations have eigvals and some don't; I thought it might be diagonal vs. non-diagonal, or things that are also observables, but that doesn't seem to be the case. @josh146 do you know the reason for this?

**Review by glassnotes — APPROVED** [review]
@mudit2812 thanks for making the requested changes, this looks good on my end :tada:

File: pennylane/ops/qutrit/__init__.py
Comment (Jaybsoni) [reply pennylane/ops/qutrit/__init__.py:35]: Cool, we can leave it as an operation for now 👍🏼, and revisit it in a later PR

File: tests/ops/qutrit/test_qutrit_non_parametric_ops.py
Comment (Jaybsoni) [reply tests/ops/qutrit/test_qutrit_non_parametric_ops.py:75]: I agree that we don't directly use it for any measurements, but I think asking for the eigenvalues of an operator is a very natural thing. Almost as natural as asking for the matrix representation of the operator. I think we should definitely support it for non-parametric operators where the result is well known (no dependance on external parameters).  

Also, with operator arithmetic, one will have the ability to combine these operators to create observables which can then be measured. We definitely make use of the Eigen values when performing these measurements.

File: pennylane/ops/qutrit/non_parametric_ops.py
Comment (Jaybsoni) [reply pennylane/ops/qutrit/non_parametric_ops.py:216]: Sure, sounds good 👍🏼

File: pennylane/ops/qutrit/non_parametric_ops.py
Comment (Jaybsoni) [reply pennylane/ops/qutrit/non_parametric_ops.py:322]: Just a note, in place inversions will be getting deprecated in the next release or two. Anything that needs to be changed here @albi3ro?

**Review by Jaybsoni — COMMENTED** [review]
Awesome work 💯! Outside of a few small suggestions I think this is good to go!

**Comment by Jaybsoni** [comment]
Happy to approve once all open comments have been resolved 👍🏼

File: tests/ops/qutrit/test_qutrit_non_parametric_ops.py
Comment (josh146) [reply tests/ops/qutrit/test_qutrit_non_parametric_ops.py:75]: I think we only ever added eigvals consistently to observables, since they are used during expectation value computation. I don't think we internally use eigvals anywhere if they are provided by a non-observable operation

File: tests/ops/qutrit/test_qutrit_non_parametric_ops.py
Comment (Jaybsoni) [inline tests/ops/qutrit/test_qutrit_non_parametric_ops.py:83]:
```suggestion
        """Tests that the TSWAP eigenvalue matches the numpy eigenvalues of the TAdd matrix"""
```

File: tests/ops/qutrit/test_qutrit_non_parametric_ops.py
Comment (Jaybsoni) [inline tests/ops/qutrit/test_qutrit_non_parametric_ops.py:83]:
```suggestion
        """Tests that the TSWAP eigenvalue matches the numpy eigenvalues of the TSWAP matrix"""
```

**Review by Jaybsoni — COMMENTED** [review]
Amazing work 💯! Glad to see it progressing! Let me know if you will need any help getting this to be fully compatible with operator arithmetic 👍🏼

**Pull request merged**
